File: http-servers/typescript/deno-oak/src/routes/params.ts
import { Router } from "@oak/oak";
import { DEFAULT_LIMIT, MAX_FILE_BYTES, NULL_BYTE, SNIFF_LEN } from "../consts/defaults.ts";
import {
  EXPECTED_FORM_CONTENT_TYPE,
  EXPECTED_MULTIPART_CONTENT_TYPE,
  FILE_NOT_FOUND,
  FILE_NOT_TEXT,
  FILE_SIZE_EXCEEDS,
  INVALID_FORM_DATA,
  INVALID_JSON_BODY,
  INVALID_MULTIPART,
  makeError,
  ONLY_TEXT_PLAIN
} from "../consts/errors.ts";

export const paramsRoutes = new Router();

paramsRoutes.get("/search", (ctx) => {
  const q = ctx.request.url.searchParams.get("q")?.trim() || "none";
  const limitStr = ctx.request.url.searchParams.get("limit")?.trim();
  const limitNum = limitStr ? Number(limitStr) : Number.NaN;
  const limit = Number.isSafeInteger(limitNum) && !limitStr?.includes(".") ? limitNum : DEFAULT_LIMIT;
  ctx.response.body = { search: q, limit };
});

paramsRoutes.get("/url/:dynamic", (ctx) => {
  const dynamic = ctx.params.dynamic;
  ctx.response.body = { dynamic };
});

paramsRoutes.get("/header", (ctx) => {
  const header = ctx.request.headers.get("X-Custom-Header")?.trim() || "none";
  ctx.response.body = { header };
});

paramsRoutes.post("/body", async (ctx) => {
  if (!ctx.request.hasBody) {
    ctx.response.status = 400;
    ctx.response.body = makeError(INVALID_JSON_BODY, "No body");
    return;
  }

  let body: unknown;
  try {
    body = await ctx.request.body.json();
  } catch (err) {
    ctx.response.status = 400;
    ctx.response.body = makeError(INVALID_JSON_BODY, err);
    return;
  }

  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    ctx.response.status = 400;
    ctx.response.body = makeError(INVALID_JSON_BODY, "expected a JSON object");
    return;
  }

  ctx.response.body = { body };
});

paramsRoutes.get("/cookie", async (ctx) => {
  const cookie = (await ctx.cookies.get("foo"))?.trim() || "none";

  await ctx.cookies.set("bar", "12345", {
    maxAge: 10,
    httpOnly: true,
    path: "/"
  });
  ctx.response.body = { cookie };
});

paramsRoutes.post("/form", async (ctx) => {
  const contentType = ctx.request.headers.get("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("application/x-www-form-urlencoded") && !contentType.startsWith("multipart/form-data")) {
    ctx.response.status = 400;
    ctx.response.body = makeError(INVALID_FORM_DATA, EXPECTED_FORM_CONTENT_TYPE);
    return;
  }

  const form: Record<string, string> = {};
  try {
    const formData = await ctx.request.body.formData();
    formData.forEach((value, key) => {
      if (typeof value === "string") {
        form[key] = value;
      }
    });
  } catch (err) {
    ctx.response.status = 400;
    ctx.response.body = makeError(INVALID_FORM_DATA, err);
    return;
  }

  const name = form.name?.trim() || "none";

  const ageStr = form.age?.trim() || "0";
  const ageNum = Number(ageStr);
  const age = Number.isSafeInteger(ageNum) ? ageNum : 0;

  ctx.response.body = { name, age };
});

paramsRoutes.post("/file", async (ctx) => {
  const contentType = ctx.request.headers.get("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("multipart/form-data")) {
    ctx.response.status = 400;
    ctx.response.body = makeError(INVALID_MULTIPART, EXPECTED_MULTIPART_CONTENT_TYPE);
    return;
  }

  let formData: FormData;
  try {
    formData = await ctx.request.body.formData();
  } catch (err) {
    ctx.response.status = 400;
    ctx.response.body = makeError(INVALID_MULTIPART, err);
    return;
  }

  const file = formData.get("file");
  if (!(file instanceof File)) {
    ctx.response.status = 400;
    ctx.response.body = makeError(FILE_NOT_FOUND, "no file field named 'file' in form data");
    return;
  }

  if (!file.type || !file.type.startsWith("text/plain")) {
    ctx.response.status = 415;
    ctx.response.body = makeError(ONLY_TEXT_PLAIN, `received mimetype: ${file.type || "unknown"}`);
    return;
  }

  if (file.size > MAX_FILE_BYTES) {
    ctx.response.status = 413;
    ctx.response.body = makeError(FILE_SIZE_EXCEEDS, `file size ${file.size} exceeds limit ${MAX_FILE_BYTES}`);
    return;
  }

  const data = new Uint8Array(await file.arrayBuffer());

  if (data.length > MAX_FILE_BYTES) {
    ctx.response.status = 413;
    ctx.response.body = makeError(FILE_SIZE_EXCEEDS, `file size ${data.length} exceeds limit ${MAX_FILE_BYTES}`);
    return;
  }

  const head = data.slice(0, SNIFF_LEN);
  if (head.includes(NULL_BYTE)) {
    ctx.response.status = 415;
    ctx.response.body = makeError(FILE_NOT_TEXT, "file contains null bytes in header");
    return;
  }

  if (data.includes(NULL_BYTE)) {
    ctx.response.status = 415;
    ctx.response.body = makeError(FILE_NOT_TEXT, "file contains null bytes");
    return;
  }

  let content: string;
  try {
    content = new TextDecoder("utf-8", { fatal: true }).decode(data);
  } catch {
    ctx.response.status = 415;
    ctx.response.body = makeError(FILE_NOT_TEXT, "file is not valid UTF-8");
    return;
  }

  ctx.response.body = {
    filename: file.name,
    size: data.length,
    content
  };
});

File: servers/py-django/src/api/views/params.py
from __future__ import annotations

import json
from typing import Any

from django.http import HttpRequest, HttpResponse, JsonResponse

from bench_shared.consts import DEFAULT_LIMIT, MAX_FILE_BYTES, NULL_BYTE, SAFE_INT_LIMIT, SNIFF_LEN
from bench_shared.errors import (
    EXPECTED_FORM_CONTENT_TYPE,
    EXPECTED_MULTIPART_CONTENT_TYPE,
    FILE_NOT_FOUND,
    FILE_NOT_TEXT,
    FILE_SIZE_EXCEEDS,
    INVALID_FORM_DATA,
    INVALID_JSON_BODY,
    INVALID_MULTIPART,
    ONLY_TEXT_PLAIN,
    make_error,
)


def _error(error: str, status: int, detail: str | Exception | None = None) -> JsonResponse:
    return JsonResponse(make_error(error, detail), status=status)


def _strip_or(value: str | None, default: str) -> str:
    stripped = value.strip() if value else ""
    return stripped or default


def _parse_safe_int(value: str | None, default: int) -> int:
    if value is None or "." in value:
        return default
    try:
        num = int(value)
    except ValueError:
        return default
    if -SAFE_INT_LIMIT <= num <= SAFE_INT_LIMIT:
        return num
    return default


async def search_params(request: HttpRequest) -> JsonResponse:
    return JsonResponse(
        {
            "search": _strip_or(request.GET.get("q"), "none"),
            "limit": _parse_safe_int(request.GET.get("limit"), DEFAULT_LIMIT),
        }
    )


async def url_params(request: HttpRequest, dynamic: str) -> JsonResponse:
    return JsonResponse({"dynamic": dynamic})


async def header_params(request: HttpRequest) -> JsonResponse:
    return JsonResponse({"header": _strip_or(request.headers.get("X-Custom-Header"), "none")})


async def body_params(request: HttpRequest) -> JsonResponse:
    try:
        payload: Any = json.loads(request.body)
    except ValueError as e:
        return _error(INVALID_JSON_BODY, 400, e)
    if not isinstance(payload, dict):
        return _error(INVALID_JSON_BODY, 400, "expected a JSON object")
    return JsonResponse({"body": payload})


async def cookie_params(request: HttpRequest) -> JsonResponse:
    response = JsonResponse({"cookie": _strip_or(request.COOKIES.get("foo"), "none")})
    response.set_cookie("bar", "12345", max_age=10, httponly=True, path="/")
    return response


async def form_params(request: HttpRequest) -> HttpResponse:
    content_type = request.headers.get("content-type", "").lower()
    if not (
        content_type.startswith("application/x-www-form-urlencoded") or content_type.startswith("multipart/form-data")
    ):
        return _error(INVALID_FORM_DATA, 400, EXPECTED_FORM_CONTENT_TYPE)

    name = _strip_or(request.POST.get("name"), "none")
    age_raw = request.POST.get("age")
    age = _parse_safe_int(age_raw if age_raw and age_raw.strip() else None, 0)
    return JsonResponse({"name": name, "age": age})


async def file_params(request: HttpRequest) -> HttpResponse:
    content_type = request.headers.get("content-type", "").lower()
    if not content_type.startswith("multipart/form-data"):
        return _error(INVALID_MULTIPART, 400, EXPECTED_MULTIPART_CONTENT_TYPE)

    upload = request.FILES.get("file")
    if upload is None:
        return _error(FILE_NOT_FOUND, 400, "no file field named 'file' in form data")

    part_type = (upload.content_type or "").lower()
    if not part_type.startswith("text/plain"):
        return _error(ONLY_TEXT_PLAIN, 415, f"received mimetype: {upload.content_type or 'unknown'}")

    data = upload.read()
    if len(data) > MAX_FILE_BYTES:
        return _error(FILE_SIZE_EXCEEDS, 413, f"file size {len(data)} exceeds limit {MAX_FILE_BYTES}")

    if NULL_BYTE in data[:SNIFF_LEN]:
        return _error(FILE_NOT_TEXT, 415, "file contains null bytes in header")
    if NULL_BYTE in data:
        return _error(FILE_NOT_TEXT, 415, "file contains null bytes")

    try:
        content = data.decode("utf-8")
    except UnicodeDecodeError as e:
        return _error(FILE_NOT_TEXT, 415, e)

    return JsonResponse({"filename": upload.name, "size": len(data), "content": content})
